File: Backend3/src/main/java/edu/cit/backend3/service/WarningService.java
package edu.cit.backend3.service;

import edu.cit.backend3.dto.CreateWarningRequest;
import edu.cit.backend3.dto.WarningDto;
import edu.cit.backend3.models.Member;
import edu.cit.backend3.models.Warning;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface WarningService {
    /**
     * Create a warning for a user
     * @param request The warning request
     * @param warnedBy The admin/moderator who issued the warning
     * @return The created warning
     */
    WarningDto createWarning(CreateWarningRequest request, Member warnedBy);
    
    /**
     * Get warnings for a specific user
     * @param memberId The ID of the user
     * @param pageable Pagination information
     * @return Page of warnings
     */
    Page<WarningDto> getWarningsForMember(Long memberId, Pageable pageable);
    
    /**
     * Get warnings issued by a specific admin/moderator
     * @param adminId The ID of the admin/moderator
     * @param pageable Pagination information
     * @return Page of warnings
     */
    Page<WarningDto> getWarningsIssuedBy(Long adminId, Pageable pageable);
    
    /**
     * Get all warnings in the system
     * @param pageable Pagination information
     * @return Page of warnings
     */
    Page<WarningDto> getAllWarnings(Pageable pageable);
    
    /**
     * Get a warning by ID
     * @param warningId The warning ID
     * @return The warning
     */
    WarningDto getWarningById(Long warningId);
    
    /**
     * Ban a user directly
     * @param memberId The ID of the user to ban
     * @param reason The reason for the ban
     * @param bannedBy The admin who banned the user
     * @return The updated member
     */
    Member banMember(Long memberId, String reason, Member bannedBy);
    
    /**
     * Unban a user
     * @param memberId The ID of the user to unban
     * @return The updated member
     */
    Member unbanMember(Long memberId);
}

File: FrontEnd/cit-forums-ui/src/services/warningService.ts
import { Warning, CreateWarningRequest } from '../models/Warning';
import axiosInstance from './axiosInstance';
import { AxiosError } from 'axios';

export const createWarning = async (request: CreateWarningRequest): Promise<Warning> => {
  try {
    console.log('Creating warning with data:', request);
    const response = await axiosInstance.post('/api/admin/warnings/create', request);
    return response.data;
  } catch (error) {
    console.error('Error in createWarning:', error);
    if (error instanceof AxiosError && error.response) {
      console.error('Error details:', error.response.data || 'No response data');
      console.error('Status code:', error.response.status);
    }
    throw error;
  }
};

export const getAllWarnings = async (page = 0, size = 10, sortBy = 'createdAt', direction = 'desc') => {
  try {
    const response = await axiosInstance.get(
      `/api/admin/warnings?page=${page}&size=${size}&sortBy=${sortBy}&direction=${direction}`
    );
    return response.data;
  } catch (error) {
    console.error('Error in getAllWarnings:', error);
    throw error;
  }
};

export const getWarningsForMember = async (memberId: number, page = 0, size = 10) => {
  try {
    const response = await axiosInstance.get(
      `/api/admin/warnings/member/${memberId}?page=${page}&size=${size}`
    );
    return response.data;
  } catch (error) {
    console.error('Error in getWarningsForMember:', error);
    throw error;
  }
};

export const getWarningById = async (warningId: number): Promise<Warning> => {
  try {
    const response = await axiosInstance.get(`/api/admin/warnings/${warningId}`);
    return response.data;
  } catch (error) {
    console.error('Error in getWarningById:', error);
    throw error;
  }
};

export const banMember = async (memberId: number, reason: string): Promise<void> => {
  try {
    await axiosInstance.post(`/api/admin/warnings/ban/${memberId}?reason=${encodeURIComponent(reason)}`);
  } catch (error) {
    console.error('Error in banMember:', error);
    throw error;
  }
};

export const unbanMember = async (memberId: number): Promise<void> => {
  try {
    await axiosInstance.post(`/api/admin/warnings/unban/${memberId}`);
  } catch (error) {
    console.error('Error in unbanMember:', error);
    throw error;
  }
};
